lib/internal/repl/experimental/history.js

```diff
@@ -0,0 +1,51 @@
+'use strict';
+
+const {
+  StringPrototypeSplit,
+} = primordials;
+
+const { exports: fs } = require('internal/fs/promises');
+const permission = require('internal/process/permission');
+const path = require('path');
+const os = require('os');
+
+module.exports = async () => {
+  let handle;
+
+  try {
+    const historyPath = path.join(os.homedir(), '.node_repl_history');
+    if (permission.isEnabled() && permission.has('fs.write', historyPath) === false) {
+      process.stdout.write('\nAccess to FileSystemWrite is restricted.\n' +
+        'REPL session history will not be persisted.\n');
+      return {
+        __proto__: null,
+        history: [],
+        writeHistory: () => false,
+      };
+    }
+
+    handle = await fs.open(historyPath, 'a+', 0o0600);
+    const data = await handle.readFile({ encoding: 'utf8' });
+    const history = StringPrototypeSplit(data, os.EOL, 1000);
```

Comment:
The following suggestion unfortunately doesn't work at the moment, but it would be beneficial to implement it (the same way we already implemented it for streams)
```suggestion
    const data = await handle.readLines().toArray();
```